sys/net/application_layer/nanocoap/link_format.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2022 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_nanocoap
+ * @{
+ *
+ * @file
+ * @brief       NanoCoAP Link Format parser
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "net/nanocoap/link_format.h"
+#include "net/nanocoap_sock.h"
+#include "net/sock/util.h"
+
+struct dir_list_ctx {
+    char *buf;
+    char *cur;
+    char *end;
+    coap_link_format_handler_t cb;
+    void *ctx;
+};
+
+static int _dirlist_cb(void *arg, size_t offset, uint8_t *buf, size_t len, int more)
+{
+    (void)offset;
+    (void)more;
+
+    struct dir_list_ctx *ctx = arg;
+
+    char *end = (char *)buf + len;
+    for (char *c = (char *)buf; c < end; ++c) {
+        if (ctx->cur) {
+            if (*c == '>' || ctx->cur == ctx->end) {
+                int res;
+                *ctx->cur = 0;
+                res = ctx->cb(ctx->buf, ctx->ctx);
+                ctx->cur = NULL;
+                if (res < 0) {
+                    return res;
+                }
+            } else {
+                *ctx->cur++ = *c;
+            }
+        } else if (*c == '<') {
+            ctx->cur = ctx->buf;
+        }
+    }
+
+    return 0;
+}
+
+int nanocoap_link_format_get(nanocoap_sock_t *sock, const char *path,
+                    coap_link_format_handler_t cb, void *arg)
+{
+    char buffer[CONFIG_NANOCOAP_QS_MAX];
```

Comment:
Hmm, a directory listing can be ***huge***. (It is disallowed to provide huge listings over CoAP+UDP without blockwise to prevent amplifying DoS attacks, but below explicitly blockwise transfer is supported.)

An alternative approach would be to parse partial data (that is until the next `,`) and receive into a ring-buffer / use `memmove()`.

Reply:
But this is just a single entry

Reply:
@maribu:
> An alternative approach would be to parse partial data (that is until the next `,`) and receive into a ring-buffer / use `memmove()`.

seems to me as if @benpicco is doing that istead of `,` to `,` going from `<`
 to `>`

Reply:
Hmm, the format is `<URI1>[;attr1[;attr2[;attr3...]]],<URI2>[;attr1[;attr2[;attr3...]]]`. And `attr1` etc. can be flags (e.g. `obs` for observe`), key-value pairs with a single value (e.g. `ct=0`), or key-value pair with multiple values (`ct=0;ct=40;`, or, as done by aiocoap IMO incompatible to the standard, `ct="0 40"`).

So it should be from `,` to `,`.

Reply:
Ok, now I get

```
</validate>;rt="validate";ct=0
</hello>;rt="Type1";ct=0;if="If1"
</bl%C3%A5b%C3%A6rsyltet%C3%B8y>;rt="blåbærsyltetøy";ct=0
</sink>;rt="sink";ct=0
</separate>;rt="separate";ct=0
</large>;rt="Type1 Type2";ct=0;sz=1700;if="If2"
</secret>;rt="secret";ct=0
</broken>;rt="Type2 Type1";ct=0;if="If2 If1"
</weird33>;rt="weird33";ct=0
</weird44>;rt="weird44";ct=0
</weird55>;rt="weird55";ct=0
</weird333>;rt="weird333";ct=0
</weird3333>;rt="weird3333";ct=0
</weird33333>;rt="weird33333";ct=0
</123412341234123412341234>;rt="123412341234123412341234";ct=0
</location-query>;rt="location-query";ct=0
</create1>;rt="create1";ct=0
</large-update>;rt="large-update";ct=0
</large-create>;rt="large-create";ct=0
</query>;rt="query";ct=0
</seg1>;rt="seg1";ct=40
</path>;rt="path";ct=40
</location1>;rt="location1";ct=40
</multi-format>;rt="multi-format";ct=0
</3>;rt="3";ct=50
</4>;rt="4";ct=50
</5>;rt="5";ct=50
```

from `coap://coap.me/`

Reply:
```
> ncget coap://coap.me/
/test
/validate
/hello
/blåbærsyltetøy
/sink
/separate
/large
/secret
/broken
/weird33
/weird44
/weird55
/weird333
/weird3333
/weird33333
/123412341234123412341234
/location-query
/create1
/large-update
/large-create
/query
/seg1
/path
/location1
/multi-format
/3
/4
/5
```